src/uniformsort.hpp

```diff
@@ -67,35 +60,35 @@ namespace UniformSort {
             buf_size--;
             // First unique bits in the entry give the expected position of it in the sorted array.
             // We take 'bucket_length' bits starting with the first unique one.
-            uint64_t pos =
-                Util::ExtractNum(buffer.get() + buf_ptr, entry_len, bits_begin, bucket_length) *
-                entry_len;
+            uint64_t idx =
+                Util::ExtractNum(buffer.get() + buf_ptr, entry_len, bits_begin, bucket_length);
+            uint64_t mem_ptr = idx * entry_len;
             // As long as position is occupied by a previous entry...
-            while (!IsPositionEmpty(memory + pos, entry_len) && pos < memory_len) {
+            while (is_used[idx] && idx < rounded_entries) {
                 // ...store there the minimum between the two and continue to push the higher one.
                 if (Util::MemCmpBits(
-                        memory + pos, buffer.get() + buf_ptr, entry_len, bits_begin) > 0) {
-                    memcpy(swap_space.get(), memory + pos, entry_len);
-                    memcpy(memory + pos, buffer.get() + buf_ptr, entry_len);
+                        memory + mem_ptr, buffer.get() + buf_ptr, entry_len, bits_begin) > 0) {
+                    memcpy(swap_space.get(), memory + mem_ptr, entry_len);
+                    memcpy(memory + mem_ptr, buffer.get() + buf_ptr, entry_len);
                     memcpy(buffer.get() + buf_ptr, swap_space.get(), entry_len);
-                    swaps++;
                 }
-                pos += entry_len;
+                idx++;
+                mem_ptr += entry_len;
             }
             // Push the entry in the first free spot.
-            memcpy(memory + pos, buffer.get() + buf_ptr, entry_len);
+            memcpy(memory + mem_ptr, buffer.get() + buf_ptr, entry_len);
+            is_used[idx] = 1;
             buf_ptr += entry_len;
         }
         uint64_t entries_written = 0;
         // Search the memory buffer for occupied entries.
-        for (uint64_t pos = 0; entries_written < num_entries && pos < memory_len;
-             pos += entry_len) {
-            if (!IsPositionEmpty(memory + pos, entry_len)) {
+        for (uint64_t idx = 0; entries_written < num_entries && idx < rounded_entries; idx++) {
+            if (is_used[idx]) {
                 // We've found an entry.
                 // write the stored entry itself.
                 memcpy(
                     memory + entries_written * entry_len,
-                    memory + pos,
+                    memory + idx * entry_len,
```

Comment:
in the loop above, you introduce a separate variable (`mem_ptr`), adding `entry_len` each iteration through the loop. I assume you did this to avoid the multiplication inside the loop. Is there a reason not to do the same thing here?

Reply:
I mostly introduced `mem_ptr` to simplify the code, I'd expect the compiler to automatically create the temporary variable instead of multiplying each time.
But yeah, making it explicit is probably a good idea.

Reply:
I would also expect the compiler to replace the multiplication in the loop. Either way is fine, I just wanted to make sure it was deliberate.